src/main/java/org/lantern/PublicIpInfoHandler.java

```diff
@@ -46,72 +42,66 @@ public PublicIpInfoHandler(final Model model, final Censored censored,
         this.geoIpLookupService = geoIpLookupService;
         Events.register(this);
     }
-    
+
+    /**
+     * We determine the public ip anytime that we've achieved internet
+     * connectivity.
+     * 
+     * @param event
+     */
+    @Subscribe
+    public void onConnectivityChanged(final ConnectivityChangedEvent event) {
+        log.debug("Got ConnectivityChangedEvent");
+        if (event.isConnected()) {
+            log.debug("Determining public ip");
+            try {
+                determinePublicIp();
+            } catch (final ConnectException e) {
+                log.warn("Could not get public IP?", e);
+            }
+        } else {
+            log.debug("Not connected to the internet, not determining public ip");
+        }
+    }
+
     /**
-     * We set the IP address on a proxy connection because we use the proxy
-     * itself to determine the IP address. This helps to minimize Lantern's
-     * network footprint.
+     * Determines the public ip address using PublicIpAddress.
      * 
-     * @throws ConnectException If there was an error fetching the public IP. 
+     * @throws ConnectException
+     *             If there was an error fetching the public IP.
      */
     private void determinePublicIp() throws ConnectException {
-        if (!this.recheck.get()) {
-            log.debug("Not rechecking for public IP");
-            return;
+        InetAddress address = null;
+        for (int i = 0; i < 10; i++) {
+            // Try a few times to get the public ip address
+            // We do this because for some reason, if we've just restored
+            // connectivity and gotten a connected event, connecting to the
+            // geo service to do an IP lookup sometimes fails.
+            try {
```

Comment:
This was a little weird, but I saw that we would get Connectivity events but then fail to connect to geoip to do the public ip lookup.  Doing this retry with exponential backoff thing seems to do the trick.